File: 2022/java/src/main/Day1.java
package main;

import java.io.*;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class Day1 {
  public static void main(String[] args) {
    try {
      List<Integer> sums = readSumsFromFile("src/resources/day1.txt");
      printSortedSums(sums);
      printTop3Sum(sums);
    } catch (Exception e) {
      System.out.println(e);
    }
  }

  public static List<Integer> readSumsFromFile(String filename) throws IOException {
    List<Integer> sums = new ArrayList<>();
    try (BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(filename)))) {
      String line;
      int sum = 0;
      while ((line = br.readLine()) != null) {
        if (!line.isEmpty()) {
          sum += parseLineAsInteger(line);
        } else {
          sums.add(sum);
          sum = 0;
        }
      }
      sums.add(sum);
    }
    return sums;
  }

  public static int parseLineAsInteger(String line) {
    try {
      return Integer.parseInt(line);
    } catch (NumberFormatException e) {
      return 0;
    }
  }

  public static void printSortedSums(List<Integer> sums) {
    sums.sort(Comparator.reverseOrder());
    System.out.println("Sorted sums: " + sums);
  }

  public static void printTop3Sum(List<Integer> sums) {
    int top3Sum = sums.stream().limit(3).mapToInt(Integer::intValue).sum();
    System.out.println("Top 3 sum: " + top3Sum);
  }
}

File: 2022/kotlin/src/main/kotlin/Day1.kt
import java.io.BufferedReader
import java.io.FileInputStream
import java.io.InputStreamReader

fun main() {
    val sums: List<Int> = readSumsFromFile("src/main/resources/day1.txt")
    printSortedSums(sums)
    printTop3Sum(sums)
}

private fun readSumsFromFile(filename: String): List<Int> {
    val sums: MutableList<Int> = ArrayList()
    BufferedReader(InputStreamReader(FileInputStream(filename))).use { br ->
        var line: String?
        var sum = 0
        while (br.readLine().also { line = it } != null) {
            if (line!!.isNotEmpty()) {
                sum += parseLineAsInteger(line!!)
            } else {
                sums.add(sum)
                sum = 0
            }
        }
        sums.add(sum)
    }
    return sums
}

private fun parseLineAsInteger(line: String): Int {
    return try {
        line.toInt()
    } catch (e: NumberFormatException) {
        0
    }
}

private fun printSortedSums(sums: List<Int>) {
    println("Sorted sums: ${sums.sortedDescending()}")
}

private fun printTop3Sum(sums: List<Int>) {
    val top3Sum = sums.sortedDescending().stream().limit(3).mapToInt { obj: Int -> obj }.sum()
    println("Top 3 sum: $top3Sum")
}
